src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -153,6 +158,34 @@ runLanguageServer options userHandlers getIdeState = do
                                 "Exception: " ++ show e
                             sendFunc $ wrap $ ResponseMessage "2.0" (responseId _id) Nothing $
                                 Just $ ResponseError InternalError (T.pack $ show e) Nothing
+                    ResponseAndRequest x@RequestMessage{_id, _params} wrap wrapNewReq act ->
+                        flip finally (clearReqId _id) $
+                        catch (do
+                            -- We could optimize this by first checking if the id
+                            -- is in the cancelled set. However, this is unlikely to be a
+                            -- bottleneck and the additional check might hide
+                            -- issues with async exceptions that need to be fixed.
+                            cancelOrRes <- race (waitForCancel _id) $ act lspFuncs ide _params
+                            case cancelOrRes of
+                                Left () -> do
+                                    logDebug (ideLogger ide) $ T.pack $
+                                        "Cancelled request " <> show _id
+                                    sendFunc $ wrap $ ResponseMessage "2.0" (responseId _id) Nothing $
+                                        Just $ ResponseError RequestCancelled "" Nothing
+                                Right (res, newReq) -> do
+                                    sendFunc $ wrap $ ResponseMessage "2.0" (responseId _id) (Just res) Nothing
+                                    case newReq of
+                                        Nothing -> return ()
+                                        Just (rm, newReqParams) -> do
+                                            reqId <- IdInt <$> randomIO
```

Comment:
This should use https://hackage.haskell.org/package/haskell-lsp-0.18.0.0/docs/Language-Haskell-LSP-Core.html#v:getNextReqId instead of generating a random request id.